### PR TITLE
Added the Salt parameter to EncryptStream + DecryptStream

### PR DESCRIPTION
**Pull request by angela2389**
Hi @PredragMaricic and @JesperSchulz,

I added the Salt to the EncryptStream + DecryptStream. Can you please review these and merge the changes to both BC15 and BC16?

I saw the test functions for these two were also missing here in Github, but I did provide them earlier to @PredragMaricic by e-mail, so they should be in your own repository. I did add the salt in the test functions as well.

### CODE REVIEW

File: Modules/System/Cryptography Management/src/DESCryptoServiceProvider.Codeunit.al [hunk 1]
```diff
@@ -47,23 +47,25 @@ codeunit 1379 DESCryptoServiceProvider
     /// Encrypts data in stream with DotNet Cryptography.DESCryptoServiceProvider
     /// </summary>
     /// <param name="Password">Represents the password to be used to initialize a new instance of Rfc2898DeriveBytes</param>
+    /// <param name="Salt">Represents the salt to be used to initialize a new instance of System.Security.Cryptography.Rfc2898DeriveBytes</param>
     /// <param name="InputInstream">Represents the input instream data to encrypt</param>
     /// <param name="OutputOutstream">Represents the output instream encrypted data</param>
     [NonDebuggable]
-    procedure EncryptStream(Password: Text; InputInstream: InStream; var OutputOutstream: Outstream)
+    procedure EncryptStream(Password: Text; Salt: Text; InputInstream: InStream; var OutputOutstream: Outstream)
```
Comment (JesperSchulz) [inline Modules/System/Cryptography Management/src/DESCryptoServiceProvider.Codeunit.al:54]: You are changing public function signatures, by adding the salt parameter. That would be a breaking change, which is no longer allowed. Please add function overloads instead. If the functions with the old signature aren't useful any longer, please deprecate those.

File: Modules/System/Cryptography Management/src/DESCryptoServiceProvider.Codeunit.al [hunk 2]
```diff
@@ -47,23 +47,25 @@ codeunit 1379 DESCryptoServiceProvider
...
     /// <summary>
     /// Decrypts data in stream with DotNet Cryptography.DESCryptoServiceProvider
     /// </summary>
     /// <param name="Password">Represents the password to be used to initialize a new instance of Rfc2898DeriveBytes</param>
+    /// <param name="Salt">Represents the salt to be used to initialize a new instance of System.Security.Cryptography.Rfc2898DeriveBytes</param>
     /// <param name="InputInstream">Represents the input instream data to decrypt</param>
     /// <param name="OutputOutstream">Represents the output instream decrypted data</param>
     [NonDebuggable]
-    procedure DecryptStream(Password: Text; InputInstream: InStream; var OutputOutstream: Outstream)
+    procedure DecryptStream(Password: Text; Salt: Text; InputInstream: InStream; var OutputOutstream: Outstream)
```
Comment (JesperSchulz) [inline Modules/System/Cryptography Management/src/DESCryptoServiceProvider.Codeunit.al:67]: Again, changed function signature.

File: Modules/System/Cryptography Management/src/DESCryptoServiceProviderImpl.Codeunit.al [hunk 3]
```diff
@@ -40,30 +40,30 @@ codeunit 1382 "DESCryptoServiceProvider Impl."
     end;
 
     [NonDebuggable]
-    procedure EncryptStream(Password: Text; InputInstream: InStream; VAR OutputOutstream: Outstream)
+    procedure EncryptStream(Password: Text; Salt: Text; InputInstream: InStream; VAR OutputOutstream: Outstream)
```
Comment (JesperSchulz) [inline Modules/System/Cryptography Management/src/DESCryptoServiceProviderImpl.Codeunit.al:43]: As this is an internal implementation codeunit, you are free to change function signatures. Just ensure, that the behavior of the function doesn't change, when called by the public facade codeunits.

**Review by JesperSchulz — CHANGES_REQUESTED** [review]
Please avoid the introduction of breaking changes.

File: Modules/System/Cryptography Management/src/DESCryptoServiceProvider.Codeunit.al (hunk 1 above)
Comment (angela2389) [reply Modules/System/Cryptography Management/src/DESCryptoServiceProvider.Codeunit.al:79]: Fair enough, they aren't useful like this, so I made them obsolete.

File: Modules/System/Cryptography Management/src/DESCryptoServiceProviderImpl.Codeunit.al (hunk 3 above)
Comment (angela2389) [reply Modules/System/Cryptography Management/src/DESCryptoServiceProviderImpl.Codeunit.al:43]: It does, they are both text parameters and when you call the new function with the "Salt" paramater you could basically fill both parameters with the same text for the function to continue working the way it did.

File: Modules/System Tests/Cryptography Management/src/DESCryptoServiceProviderTest.Codeunit.al [hunk 4]
```diff
@@ -46,4 +46,77 @@ codeunit 132588 "DESCryptoServiceProvider Test"
...
+        DESCryptoServiceProvider.EncryptStreamWithDESCryptoServiceProvider(Password, Salt, InputInstream, OutputOutstream);
+
+        TempBlob.Blob.CreateInStream(OutputInstream);
+        EncryptedStreamText := Base64Convert.ToBase64(OutputInstream);
+
+        // [THEN] Verify Result 
+        LibraryAssert.AreEqual(ExpectedEncryptedText, EncryptedStreamText, 'Unexpected value when decrypting stream using DESCryptoServiceProvider');
+    end;
+
+    procedure TestDecryptStream()
```
Comment (stkillen) [inline Modules/System Tests/Cryptography Management/src/DESCryptoServiceProviderTest.Codeunit.al:85]: Same question as above.

```suggestion
    [Test]
    procedure TestDecryptStream()
```

File: Modules/System/Cryptography Management/src/DESCryptoServiceProvider.Codeunit.al [hunk 5]
```diff
@@ -42,14 +42,14 @@ codeunit 1379 DESCryptoServiceProvider
...
     /// Encrypts data in stream with DotNet Cryptography.DESCryptoServiceProvider
     /// </summary>
     /// <param name="Password">Represents the password to be used to initialize a new instance of Rfc2898DeriveBytes</param>
     /// <param name="InputInstream">Represents the input instream data to encrypt</param>
     /// <param name="OutputOutstream">Represents the output instream encrypted data</param>
     [NonDebuggable]
+    [Obsolete('(Removed,Add the salt parameter to continue using this function)')]
     procedure EncryptStream(Password: Text; InputInstream: InStream; var OutputOutstream: Outstream)
     begin
         DESCryptoServiceProviderImpl.EncryptStream(Password, InputInstream, OutputOutstream);
```
Comment (stkillen) [inline Modules/System/Cryptography Management/src/DESCryptoServiceProvider.Codeunit.al:55]: With the signature change in the implementation codeunit, this call is now missing the `Salt `parameter. You should use the password as the salt to ensure the same behavior as before.

```suggestion
        DESCryptoServiceProviderImpl.EncryptStream(Password, Password, InputInstream, OutputOutstream);
```

File: Modules/System/Cryptography Management/src/DESCryptoServiceProvider.Codeunit.al [hunk 6]
```diff
@@ -62,8 +62,35 @@ codeunit 1379 DESCryptoServiceProvider
     /// <param name="InputInstream">Represents the input instream data to decrypt</param>
     /// <param name="OutputOutstream">Represents the output instream decrypted data</param>
     [NonDebuggable]
+    [Obsolete('(Removed,Add the salt parameter to continue using this function)')]
     procedure DecryptStream(Password: Text; InputInstream: InStream; var OutputOutstream: Outstream)
     begin
         DESCryptoServiceProviderImpl.DecryptStream(Password, InputInstream, OutputOutstream);
```
Comment (stkillen) [inline Modules/System/Cryptography Management/src/DESCryptoServiceProvider.Codeunit.al:68]: Same as above.

```suggestion
        DESCryptoServiceProviderImpl.DecryptStream(Password, Password, InputInstream, OutputOutstream);
```

**Review by stkillen — CHANGES_REQUESTED** [review]
Hi @angela2389, just a few comments

**Review by stkillen — APPROVED** [review]

**Review by JesperSchulz — APPROVED** [review]

**Comment by bc-ghost** [comment]
Thanks for submitting this issue. We will publish a fix for it in the next minor update for release 17.

**Comment by bc-ghost** [comment]
Thanks for reporting this. We agree, and we’ll publish a fix asap, either in an update for the current version or in the next major release. We will update this issue with information about availability.

**Comment by bc-ghost** [comment]
Thanks for submitting this issue. We will publish a fix for it in the next minor update for release 16.

**Comment by JesperSchulz** [comment]
Availability update: We will publish a fix for this issue in the next update for releasee 15.

**Pull request merged**